tiatoolbox/wsicore/wsireader.py

```
@@ -103,52 +101,68 @@ def open(
+            )
+            return is_dcm or is_dcm_dir
 
-            if suffixes[-1] in (".jp2",):
-                return OmnyxJP2WSIReader(input_img, mpp=mpp, power=power)
+        if not isinstance(input_img, (WSIReader, np.ndarray, str, pathlib.Path)):
+            raise TypeError(
+                "Invalid input: Must be a WSIRead, numpy array, string or pathlib.Path"
+            )
 
         if isinstance(input_img, np.ndarray):
             return VirtualWSIReader(input_img, mpp=mpp, power=power)
 
-        if isinstance(
-            input_img,
-            WSIReader,
-        ):
-            # input is already a tiatoolbox wsi handler
+        if isinstance(input_img, WSIReader):
             return input_img
-        raise TypeError("Invalid input. Must be a file path or an ndarray image.")
+
+        if is_dicom(input_img):
+            return DICOMWSIReader(input_img, mpp=mpp, power=power)
+
+        _, _, suffixes = utils.misc.split_path_name_ext(input_img)
+
+        if suffixes[-1] not in [
+            ".svs",
+            ".npy",
+            ".ndpi",
+            ".mrxs",
+            ".tif",
+            ".tiff",
+            ".jp2",
+            ".png",
+            ".jpg",
+            ".jpeg",
+        ]:
+            raise FileNotSupported(f"File {input_img} is not a supported file format.")
+
+        if suffixes[-1] in (".npy",):
+            input_img = np.load(input_img)
+            return VirtualWSIReader(input_img, mpp=mpp, power=power)
+
+        if suffixes[-2:] in ([".ome", ".tiff"],):
+            return TIFFWSIReader(input_img, mpp=mpp, power=power)
+
+        if suffixes[-1] in (".jpg", ".jpeg", ".png", ".tif", ".tiff"):
+            return VirtualWSIReader(input_img, mpp=mpp, power=power)
+
+        if suffixes[-1] in (".jp2",):
+            return OmnyxJP2WSIReader(input_img, mpp=mpp, power=power)
+
+        return OpenSlideWSIReader(input_img, mpp=mpp, power=power)
 
```

just following up on previous suggestion
you can delete lines 138 - 149 and raise after this return statement.

---

This will cause a linter issue

---

same as above, please see https://github.com/TissueImageAnalytics/tiatoolbox/blob/47b949421a6a942df593bea104650999738acb87/tiatoolbox/utils/misc.py#L383

---

I do quite like the raise at the end pattern, as it gives a nice 'catch everything that gets through' feeling. However, it does make more sense to fail early in general, as it gives faster user feedback and avoids unnecessary processing.

---

I am happy with both, it just reduces the number of lines